File: java/572.subtree-of-another-tree.java
/*
 * @lc app=leetcode id=572 lang=java
 *
 * [572] Subtree of Another Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if(t == null)
            return true;
        if(s ==  null)
            return false;
        return isTheSame(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }
    boolean isTheSame(TreeNode a, TreeNode b){
        if(b == null && a == null)
            return true;
        if(b == null || a == null)
            return false;
        if(a.val != b.val)
            return false;
        return isTheSame(a.left, b.left) && isTheSame(a.right, b.right);
    }
}
// @lc code=end

File: unsorted/js/572.subtree-of-another-tree.js
/*
 * @lc app=leetcode id=572 lang=javascript
 *
 * [572] Subtree of Another Tree
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 * 
 */

// two layer of dfs
var isSubtree = function(s, t) {
	// if(!s) return !t;

	if (s === null) return t === null;
	else if (isEqual(s, t)) return true;
	else if (isSubtree(s.left, t)) return true;
	else if (isSubtree(s.right, t)) return true;
	else return false;
};

const isEqual = (a, b) => {
	if (a === null) return b === null;
	else if (b === null) return false;
	else if (a.val === b.val) return isEqual(a.left, b.left) && isEqual(a.right, b.right);
	else return false;
};
